### PR TITLE
Feature/update kotlin dependecies template

### PR DESCRIPTION
**Pull request by Marek00Malik**
According to #11994, this change is updating Kotlin client dependencies, Kotlin version and what version of Ktor is used (2.0.3). 

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [@jimschubert, @dr4ke616  @karismann  @Zomzog @andrewemery @4brunu  @yutaka0m ]

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/build.gradle.kts.mustache [hunk 1]
```diff
@@ -8,10 +8,10 @@ plugins {
...
 version = "{{artifactVersion}}"
 
-val kotlin_version = "1.6.0"
-val coroutines_version = "1.5.2"
-val serialization_version = "1.3.0"
-val ktor_version = "1.6.4"
+val kotlin_version = "1.6.10"
+val coroutines_version = "1.6.3"
+val serialization_version = "1.3.3"
+val ktor_version = "2.0.3"
```
[the remark this answers is not in the document]
Comment (Marek00Malik) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/build.gradle.kts.mustache:14]: There are some breaking changes in the 2.x release yes. But it depends on how much we are using the ktor library.

[the remark this answers is not in the document]
Comment (4brunu) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/build.gradle.kts.mustache:14]: Can you please try to build the Multiplatform sample project to make sure it works please? Thanks

[the remark this answers is not in the document]
Comment (Marek00Malik) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/build.gradle.kts.mustache:14]: @4brunu should be fixed. Thanks for the hints again.

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache [hunk 2]
```diff
@@ -199,13 +227,3 @@ import {{packageName}}.auth.*
...
-
-{{#gson}}private class JsonSerializerImpl(private val gson: Gson) : JsonSerializer {
-  override fun write(data: Any, contentType: ContentType): OutgoingContent =
-    ByteArrayContent(gson.toJson(data).toByteArray(StandardCharsets.UTF_8), contentType)
-}{{/gson}}
-
-{{#jackson}}private class JsonSerializerImpl(private val objectMapper: ObjectMapper) : JsonSerializer {
-  override fun write(data: Any, contentType: ContentType): OutgoingContent =
-    ByteArrayContent(objectMapper.writeValueAsBytes(data), contentType)
-}{{/jackson}}
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache:211]: @Marek00Malik this is not necessary anymore?

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache [hunk 3]
```diff
@@ -148,18 +182,13 @@ import {{packageName}}.auth.*
...
-    protected suspend fun <T: Any?> jsonRequest(requestConfig: RequestConfig<T>, body: Any? = null, authNames: kotlin.collections.List<String>): HttpResponse {
-        val contentType = (requestConfig.headers[HttpHeaders.ContentType]?.let { ContentType.parse(it) }
-                ?: ContentType.Application.Json)
-        return if (body != null) request(requestConfig, serializer.write(body, contentType), authNames)
-        else request(requestConfig, authNames = authNames)
-    }
+    protected suspend fun <T: Any?> jsonRequest(requestConfig: RequestConfig<T>, body: Any? = null, authNames: kotlin.collections.List<String>): HttpResponse = request(requestConfig, body, authNames)
 
-    protected suspend fun <T: Any?> request(requestConfig: RequestConfig<T>, body: OutgoingContent = EmptyContent, authNames: kotlin.collections.List<String>): HttpResponse {
+    protected suspend fun <T: Any?> request(requestConfig: RequestConfig<T>, body: Any? = null, authNames: kotlin.collections.List<String>): HttpResponse {
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache:187]: @Marek00Malik why did you changed the body to `Any`?

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache [hunk 4]
```diff
@@ -148,18 +182,13 @@ import {{packageName}}.auth.*
         return request(requestConfig, FormDataContent(body ?: Parameters.Empty), authNames)
     }
 
-    protected suspend fun <T: Any?> jsonRequest(requestConfig: RequestConfig<T>, body: Any? = null, authNames: kotlin.collections.List<String>): HttpResponse {
-        val contentType = (requestConfig.headers[HttpHeaders.ContentType]?.let { ContentType.parse(it) }
-                ?: ContentType.Application.Json)
-        return if (body != null) request(requestConfig, serializer.write(body, contentType), authNames)
-        else request(requestConfig, authNames = authNames)
-    }
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache:156]: @Marek00Malik This code is not necessary anymore?

**Review by 4brunu — COMMENTED** [review]
Hey @Marek00Malik, I left some questions regarding the new changes.

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache (hunk 2 above)
Comment (Marek00Malik) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache:211]: @4brunu with Ktor 2.x.y introduced a new serialization plugin called ContentNegotiation. The client is responsible for the correct conversion to the type specified by the content type. 
https://ktor.io/docs/serialization-client.html

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache (hunk 3 above)
Comment (Marek00Malik) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache:187]: the new api is now asking for the actual object that should be serialized during the request. That's why we don't need OutgoingContent, but any object we are passing. I was thinking of taking advantage but couldn't find the T that would be used with the specific call so it's Any.

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache (hunk 4 above)
Comment (Marek00Malik) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache:156]: Yes, the client is now serializing the object, it's not waiting for a serialized string payload.

**Comment by 4brunu** [comment]
Can you please fix the conflicts with the master branch?

```
./mvnw clean package 
./bin/generate-samples.sh ./bin/generate-samples.sh bin/configs/kotlin*
```

**Comment by mCzolko** [comment]
Excelent work! I'm looking forward for this. Is it real updated Ktor will in in release 6.1.0?

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/infrastructure/ApiClient.kt.mustache [hunk 5]
```diff
@@ -3,9 +3,8 @@ package {{packageName}}.infrastructure
...
 import io.ktor.client.HttpClientConfig
 import io.ktor.client.engine.HttpClientEngine
-import io.ktor.client.features.json.JsonFeature
-import io.ktor.client.features.json.JsonSerializer
-import io.ktor.client.features.json.serializer.KotlinxSerializer
+import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
+import io.ktor.serialization.kotlinx.json.*
 import io.ktor.client.request.*
 import io.ktor.client.request.forms.FormDataContent
 import io.ktor.client.request.forms.MultiPartFormDataContent
```
Comment (mCzolko) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/infrastructure/ApiClient.kt.mustache:10]: unused imports:
```kt
import io.ktor.client.utils.EmptyContent
import io.ktor.http.content.OutgoingContent
import {{apiPackage}}.*
import {{modelPackage}}.*
```

Comment (Marek00Malik) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/infrastructure/ApiClient.kt.mustache:10]: done

**Comment by Marek00Malik** [comment]
@4brunu was trying to update the samples, but for some reason I'm getting this error. 
```
[main] INFO  o.o.codegen.cmd.GenerateBatch - Batch generation using up to 20 threads.
Includes: /Users/marek.macbook/DevTools/GitCode/CodeHouse/openapi-generator
Root: /Users/marek.macbook/DevTools/GitCode/CodeHouse/openapi-generator
[main] ERROR o.o.c.config.CodegenConfigurator - Unexpected character ('#' (code 35)): expected a valid value (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (File); line: 1, column: 2]
[main] ERROR o.o.c.config.CodegenConfigurator - Unexpected character ('#' (code 35)): expected a valid value (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (File); line: 1, column: 2]
Exception in thread "main" java.lang.RuntimeException: Unable to deserialize config file: ./bin/generate-samples.sh

```

This is the same for master as well.

**Comment by Marek00Malik** [comment]
@4brunu updated to master and updated sample templest

**Comment by scottkennedy** [comment]
Would it be possible to get this merged in?

**Comment by wing328** [comment]
CI failure not related to the change: https://github.com/OpenAPITools/openapi-generator/runs/8072408542?check_suite_focus=true

**Comment by frichez** [comment]
Hi @Marek00Malik I'using multiplatform library and now due to this change I'm not more able to send a request in json format using negotiation/json serialization plugin.
Previously jsonRequest() method was setting the content-type header to json. It is not more the case and the request is not serialized : "If you expect serialized body, please check that you have installed the corresponding plugin(like `ContentNegotiation`) and **set `Content-Type` header**." 
Do I miss something ? Thks.

**Comment by Marek00Malik** [comment]
Hey @frichez, in general, this new plugin should be able to send any request with application/json as the content type. 
Unfortunately, I'm not able to sit on this and add the header by default as I'm low on free time now (production release coming soon).

**Comment by frichez** [comment]
Hi @Marek00Malik, this for the answer. You can have a look in jsonRequest(), `application/json` header was set there before this change (multiplatform/infrastructure/ApiClient.kt.mustache). I'm using grade to build my project so I tried to modify ApiClient.kt.mustache with the content of the previous jsonRequest() method. Unfortunately setting templateDir in openApiGenerate grade section has not effect :-(, gradle complains if the template dir doesn't exist but don't use it as custom template dir. As fallback I had to used the generator as command file. I see an issue for [this](https://github.com/OpenAPITools/openapi-generator/issues/5463) created in 2020...

**Comment by Marek00Malik** [comment]
@frichez I've pushed a PR that adds this content type header. Please review if. If you would find any improvement feel free to contribute.

**Comment by Marek00Malik** [comment]
@frichez so I've checked in the source code how the serializer is being used I noticed that in the multiplatform case the json serializer is using the json contentType. 

```
    private val clientConfig: (HttpClientConfig<*>) -> Unit by lazy {
        {
            it.install(ContentNegotiation) { json(jsonBlock) }
            httpClientConfig?.invoke(it)
        }
    }
```
```
@OptIn(ExperimentalSerializationApi::class)
public fun Configuration.json(
    json: Json = DefaultJson,
    contentType: ContentType = ContentType.Application.Json
) {
    serialization(contentType, json)
}
```

**Comment by frichez** [comment]
> @OptIn(ExperimentalSerializationApi::class)
> public fun Configuration.json(
>     json: Json = DefaultJson,
>     contentType: ContentType = ContentType.Application.Json
> ) {
>     serialization(contentType, json)
> }

Thks @Marek00Malik. Do I need to add some code in my client or is it an issue in openapi generator ?

**Comment by Marek00Malik** [comment]
Hey @frichez from the examples that are in the repo you can see you don't need to add anything except what is in the dependencies. See this [example](https://github.com/OpenAPITools/openapi-generator/blob/master/samples/client/petstore/kotlin-multiplatform/build.gradle.kts) that is used for validating the multiplatform setup. 
If you have an issue, a sample project would be handy to verify your issue.

**Comment by frichez** [comment]
hi @Marek00Malik I tried the example you mentioned but I contains no unit tests.
I added following code in samples/client/petstore/kotlin-multiplatform/src/jvmTest/kotlin/util/PetStoreTest.kt and I get the same error as I have in my project:

If you expect serialized body, please check that you have installed the corresponding plugin(like `ContentNegotiation`) and set `Content-Type` header.
java.lang.IllegalStateException: Fail to prepare request body for sending. 
The body type is: class org.openapitools.client.models.Pet (Kotlin reflection is not available), with Content-Type: **null**.


```
package util

import org.junit.Test
import org.openapitools.client.apis.PetApi
import org.openapitools.client.models.Pet

class PetStoreTest {
    @Test
    fun addPet() = runTest {
        val mainApi = PetApi()
        println(mainApi.addPet(Pet("olya", emptyList())).body())
    }
}
```

**Comment by Marek00Malik** [comment]
@frichez 
https://github.com/OpenAPITools/openapi-generator/pull/14045

**Pull request merged**